src/input/shared.rs

```diff
@@ -74,3 +74,31 @@ pub fn float_as_int(input: &dyn Input, float: f64) -> ValResult<i64> {
         Ok(float as i64)
     }
 }
+
+/// This is around 40x faster than using `str::parse::<i64>()`
+pub fn fast_parse(s: &str) -> Option<i64> {
+    let mut bytes = s.bytes();
+    let mut result: u64 = 0;
+    let sign: i64 = match bytes.next() {
```

Comment:
It is not necessary for sign to be i64, it could be boolean instead. Also, there is atoi which could be helpful and the impl could be extracted from there. https://github.com/pacman82/atoi-rs/blob/43e1b8114ceea167679a923c9873265e3bf4b417/src/lib.rs#L371

Reply:
I didn't :( Though, it would be minimal indeed - maybe a bit fewer instructions, but I need to check deeper.
I am also wondering if you can share the benchmark where you observed 40x improvement?

Locally I observed this function to be slightly slower than stdlib on some arbitrary input:

```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};

pub fn fast_parse(s: &str) -> Option<i64> {
    let mut bytes = s.bytes();
    let mut result: u64 = 0;
    let sign: i64 = match bytes.next() {
        Some(b'-') => -1,
        Some(c) if (b'0'..=b'9').contains(&c) => {
            result = (c & 0x0f) as u64;
            1
        }
        _ => return None,
    };

    for digit in bytes {
        match digit {
            b'0'..=b'9' => {
                result *= 10;
                result += (digit & 0x0f) as u64;
                if result >= i64::MAX as u64 {
                    return None;
                }
            }
            _ => return None,
        }
    }
    Some(sign * (result as i64))
}

fn bench(c: &mut Criterion) {
    let positive = black_box("123456789");
    let negative = black_box("-123456789");
    c.bench_function("std-positive", |b| b.iter(|| str::parse::<i64>(positive)));
    c.bench_function("std-negative", |b| b.iter(|| str::parse::<i64>(negative)));
    c.bench_function("fast-positive", |b| b.iter(|| fast_parse(positive)));
    c.bench_function("fast-negative", |b| b.iter(|| fast_parse(negative)));
}

criterion_group!(arbitrary, bench);
criterion_main!(arbitrary);
```

```
rust-experiments on  master [!+?] is 📦 v0.1.0 via 🦀 v1.61.0 took 46s 
❯ cargo bench
    Finished bench [optimized] target(s) in 0.03s
     Running unittests src/lib.rs (target/release/deps/rust_experiments-5796d4cfad1eec85)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/bench.rs (target/release/deps/bench-0eff06ed3d8e42dc)
WARNING: HTML report generation will become a non-default optional feature in Criterion.rs 0.4.0.
This feature is being moved to cargo-criterion (https://github.com/bheisler/cargo-criterion) and will be optional in a future version of Criterion.rs. To silence this warning, either switch to cargo-criterion or enable the 'html_reports' feature in your Cargo.toml.

Gnuplot not found, using plotters backend
std-positive            time:   [8.3551 ns 8.4134 ns 8.4980 ns]                          
                        change: [+1.2060% +1.6968% +2.3684%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe

std-negative            time:   [8.4928 ns 8.5056 ns 8.5218 ns]                          
                        change: [+2.0719% +2.3850% +2.7353%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe

fast-positive           time:   [10.675 ns 10.680 ns 10.685 ns]                           
                        change: [-0.0221% +0.0440% +0.1025%] (p = 0.18 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low severe
  2 (2.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe

fast-negative           time:   [11.930 ns 11.942 ns 11.953 ns]                           
                        change: [-0.0077% +0.1757% +0.3382%] (p = 0.05 > 0.05)
                        No change in performance detected.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```

Reply:
Ok, my mistake, it seems that the speed up is more like 2x, which is still worthwhile but much less stark than the 40x I saw before. I think the error was because I forgot to use `black_box` 🙈 . Sorry for the confusion.

https://github.com/samuelcolvin/bench-fast-parse

Reply:
not sure about the difference with your results. Maybe arch, maybe version?

I'm using mac with m1 chip and `rustc 1.63.0-nightly (e09449220 2022-05-31)`

Reply:
see also https://rust-malaysia.github.io/code/2020/07/11/faster-integer-parsing.html for some context.